src/dune_rules/lib.ml

```diff
@@ -1248,8 +1249,27 @@ end = struct
           | Re_export lib ->
             let+ lib = resolve_dep db lib ~private_deps in
             add_re_exports acc lib
-          | Direct lib ->
+          | Direct ((lib_dep_loc, _) as lib) ->
             let+ lib = resolve_dep db lib ~private_deps in
+            let lib =
+              match modes with
+              | None -> lib
+              | Some (modes : Lib_mode.Map.Set.t) ->
+                Resolve.bind lib ~f:(fun lib ->
```

Comment:
We shouldn't fail to resolve a library just because it fails for a particular mode. The failure should be for that mode only.